### PR TITLE
chore: Anvil on canvas UI tests

### LINKED ISSUE #33979: Task: Tests for onCanvasUI
Unit tests: 
- Test AnvilWidgetNameComponent component for state based style changes
  - component styles when selected
  - component styles when hovered
  - does parent widget selector show up based on widget selected
  - check error indicator ux

Integration tests:
- implement interaction based test cases out of test plan https://www.notion.so/appsmith/Widget-highlight-and-selection-29fbaa6624854998b52e62b53a4f572e
  - mouse over ineraction to focus widget
  - click on widget to select and show widget name
  - click on parent widget selector to select parent of widget
  - drag and drop single widget via the widget name component
  - drag and drop multiple widgets via the widget name component

### PR DESCRIPTION
**Pull request by riodeuno**
### Unit
`index.test.tsx` - (Main connected component that uses floating ui)
- [x] Test rendering of widget name component (floating-ui) based on "select", "focus" and "none" states from redux selector
- [x] Test rendering of widget name component (floating-ui) based on error count from redux selector

`AnvilWidgetNameComponent.test.tsx` (Standalone component that renders content for floating ui to use)
- [x] Test rendering of standalone widget name component based on props.
- [x] Test calling of `onDragStart` callback from the standalone widget name component.

`SplitButton.test.tsx` (Low level UI component - child of `AnvilWidgetNameComponent`)
- [x] Test left and right toggle click triggers
- [x] Test if component handles props correctly

### E2E
- [x] Test Dragging via widget name component
- [x] Test widget name component when multiple widgets are selected
- [x] Test Hover and Selection to show widget name component

Fixes #33979

## Automation

/ok-to-test tags="@tag.All"

### :mag: Cypress test results

> [!TIP]
> 🟢 🟢 🟢 All cypress tests have passed! 🎉 🎉 🎉
> Workflow run: <https://github.com/appsmithorg/appsmith/actions/runs/10349986042>
> Commit: 8c9e6b933fc63306ca4b9322d2cc109909df879d
> <a href="https://internal.appsmith.com/app/cypress-dashboard/rundetails-65890b3c81d7400d08fa9ee5?branch=master&workflowId=10349986042&attempt=1" target="_blank">Cypress dashboard</a>.
> Tags: `@tag.All`
> Spec:
> <hr>Mon, 12 Aug 2024 11:13:52 UTC

## Communication
Should the DevRel and Marketing teams inform users about this change?
- [ ] Yes
- [x] No

## Summary by CodeRabbit

- **New Features**
	- Introduced a comprehensive suite of end-to-end tests for the Anvil editor's UI in canvas layout mode.
	- Enhanced flexibility in widget manipulation on the canvas with an updated method.
	- Expanded selectors for more precise targeting of UI components.
	- Added unit tests for the `AnvilWidgetNameComponent` and `SplitButton` components to ensure expected functionality.

- **Bug Fixes**
	- Improved responsiveness of the `AnvilWidgetName` component to state changes.

- **Documentation**
	- Updated comments and structure for clarity in test scenarios and component functionality.

### CODE REVIEW

**Comment by riodeuno** [comment]
@sagar-qa007 @ApekshaBhosale Please review the cypress test. I've followed some existing patterns used by existing tests. No new waits. Individual tests cleanup in the tests themselves, making them idempotent.

File: app/client/cypress/e2e/Regression/ClientSide/Anvil/AnvilOnCanvasUI_spec.ts [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+      // intercept features call for Anvil + WDS tests
+      featureFlagIntercept({
+        release_anvil_enabled: true,
+      });
+    });
+    it("1. Widget name component shows and positioned correctly", () => {
+      // Cleanup the canvas before each test
+      agHelper.PressEscape();
+      agHelper.SelectAllWidgets();
+      agHelper.PressDelete();
```
Comment (ApekshaBhosale) [inline app/client/cypress/e2e/Regression/ClientSide/Anvil/AnvilOnCanvasUI_spec.ts:21]: you can move these 3 lines into beforeEach. it's a repeated code and looks like set up for a test to run @riodeuno

**Review by ApekshaBhosale — CHANGES_REQUESTED** [review]

File: app/client/cypress/e2e/Regression/ClientSide/Anvil/AnvilOnCanvasUI_spec.ts (hunk 1 above)
Comment (riodeuno) [reply app/client/cypress/e2e/Regression/ClientSide/Anvil/AnvilOnCanvasUI_spec.ts:21]: @ApekshaBhosale This is actually clean up code for the second test. The first test doesn't really need it, but I've added it anyway.

File: app/client/cypress/e2e/Regression/ClientSide/Anvil/AnvilOnCanvasUI_spec.ts [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+
+      // DnD Switch widget
+      anvilLayout.dnd.DragDropNewAnvilWidgetNVerify(
+        anvilLocators.WDSSWITCH,
+        5,
+        20,
```
[the remark this answers is not in the document]
Comment (riodeuno) [reply app/client/cypress/e2e/Regression/ClientSide/Anvil/AnvilOnCanvasUI_spec.ts:94]: @KelvinOm These are XY distances. A standard argument for any drag and drop function. Would you like for me to comment on this code?

File: app/client/cypress/e2e/Regression/ClientSide/Anvil/AnvilOnCanvasUI_spec.ts (hunk 1 above)
Comment (riodeuno) [reply app/client/cypress/e2e/Regression/ClientSide/Anvil/AnvilOnCanvasUI_spec.ts]: @ApekshaBhosale I've removed the cleanup for the first test as it was not necessary.

File: app/client/cypress/e2e/Regression/ClientSide/Anvil/AnvilOnCanvasUI_spec.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (KelvinOm) [reply app/client/cypress/e2e/Regression/ClientSide/Anvil/AnvilOnCanvasUI_spec.ts:94]: @riodeuno I would prefer to have variables to make it easier to read. Without going into the depth of the function, it is very difficult to understand what it is. 

It is much easier to understand and read something like this. 
```
anvilLayout.dnd.DragDropNewAnvilWidgetNVerify(
        anvilLocators.WDSBUTTON,
        mouseX,
        mouseY,
        {
          skipWidgetSearch: true,
        },
      );
```

Anyway, it's a nitpick and it's up to you to fix it or not.

[the remark this answers is not in the document]
Comment (riodeuno) [reply app/client/cypress/e2e/Regression/ClientSide/Anvil/AnvilOnCanvasUI_spec.ts:94]: @KelvinOm Sure, I'll make this change in my particular test to make it readable.

**Review by KelvinOm — DISMISSED** [review]

File: app/client/cypress/e2e/Regression/ClientSide/Anvil/AnvilOnCanvasUI_spec.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (riodeuno) [reply app/client/cypress/e2e/Regression/ClientSide/Anvil/AnvilOnCanvasUI_spec.ts]: @KelvinOm I've updated the code. Also, added comments on the original function to help with the usage.

**Review by KelvinOm — APPROVED** [review]

**Comment by ApekshaBhosale** [comment]
changes look good to me @riodeuno Have you run the nest test on EE as well?

**Review by ApekshaBhosale — APPROVED** [review]

**Comment by riodeuno** [comment]
@ApekshaBhosale Ran it 5 times (1 using all tests, 4 using limited test) on EE. All runs passed. I'm merging this in.

**Pull request merged**